gcloud/storage/batch.py

```diff
@@ -71,6 +73,76 @@ class NoContent(object):
     status = 204
 
 
+class _FutureDict(object):
+    """Class to hold a future value for a deferred request.
+
+    Used by for requests that get sent in a :class:`Batch`.
+
+    :type owner: object
+    :param owner: Optional. A blob or bucket (or other type) that depends on
+                  this future.
+    """
+
+    def __init__(self, owner=None):
+        self._value = None
+        self.owner = owner
+
+    @staticmethod
+    def get(key, default=None):
+        """Stand-in for dict.get.
+
+        :type key: object
+        :param key: Hashable dictionary key.
+
+        :type default: object
+        :param default: Fallback value to dict.get.
+
+        :raises: :class:`KeyError` always since the future is intended to fail
+                 as a dictionary.
+        """
+        raise KeyError('Cannot get(%r, default=%r) on a future' % (
+            key, default))
+
+    def __getitem__(self, key):
+        """Stand-in for dict[key].
+
+        :type key: object
+        :param key: Hashable dictionary key.
+
+        :raises: :class:`KeyError` always since the future is intended to fail
+                 as a dictionary.
+        """
+        raise KeyError('Cannot get item %r from a future' % (key,))
+
+    def __setitem__(self, key, value):
+        """Stand-in for dict[key] = value.
+
+        :type key: object
+        :param key: Hashable dictionary key.
+
+        :type value: object
+        :param value: Dictionary value.
+
+        :raises: :class:`KeyError` always since the future is intended to fail
+                 as a dictionary.
+        """
+        raise KeyError('Cannot set %r -> %r on a future' % (key, value))
+
+    def set_future_value(self, value):
+        """Sets the value associated with the future.
+
+        :type value: object
+        :param value: The future value that was set.
+
+        :raises: :class:`ValueError` if the value is already set.
+        """
+        if self._value is not None:
+            raise ValueError('Future value is already set.')
+        self._value = value
+        if self.owner is not None:
+            self.owner._properties = value
+
```

Comment:
@tseaver comment from f2a9c7dfa5f753fc007a8cea4741aa99131da5bb:

> This feels wrong:  at the time we are assigning values, we should just replace the `_FutureDict` with an real dict.

----

It feels wrong to me too.

This does get called at the time we are assigning values in `Batch.finish`. I'm not clear what you mean here.